webtransport/README.md

```diff
@@ -0,0 +1,62 @@
+Examples:
+* `/ip4/1.2.3.4/udp/443/quic/webtransport/certhash/<hash1>`
+* `/ip6/fe80::1ff:fe23:4567:890a/udp/1234/quic/webtransport/certhash/<hash1>/certhash/<hash2>/certhash/<hash3>`
+
+## WebTransport HTTP endpoint
+
+WebTransport needs a HTTPS URL to establish a WebTransport session, e.g. `https://example.com/webtransport`. As multiaddrs don't allow the encoding of URLs, this spec standardizes the endpoint. The HTTP endpoint of a libp2p WebTransport servers MUST be located at `/.well-known/libp2p-webtransport`.
+
+To allow future evolution of the way we run the libp2p handshake over WebTransport, we use a URL parameter. The handshake described in this document MUST be signaled by setting the `type` URL parameter to `noise`.
+
+Example: The WebTransport URL of a WebTransport server advertising `/ip4/1.2.3.4/udp/1443/quic/webtransport/` would be `https://1.2.3.4:1443/.well-known/libp2p-webtransport?type=noise`.
+
+## Security Handshake
+
+Unfortunately, the self-signed certificate doesn't allow the nodes to authenticate each others' peer IDs. It is therefore necessary to run an additional libp2p handshake on a newly established WebTransport connection.
+The first stream that the client opens on a new WebTransport session is used to perform a libp2p handshake using Noise (https://github.com/libp2p/specs/tree/master/noise). The client SHOULD start the handshake right after sending the CONNECT request, without waiting for the server's response.
+
+In order to verify that the end-to-end encryption of the connection, the peers need to establish that no MITM intercepted the connection. To do so, the client MUST include the certificate hash that was used to establish the connection as payload of the first Noise message (the `e` message). This payload is not encrypted, but the Noise handshake provides integrity protection.
+If the client was willing to accept multiple certificate hashes, but cannot determine which certificate was actually used to establish the connection (this will commonly be the case for browser clients), it MUST include a list of all certificate hashes.
+
+Certificate hashes are encoded using the following protobuf message:
+```proto
+syntax = "proto2";
+
+message WebTransport {
+  repeated bytes cert_hashes = 1;
+}
+```
+
+On receipt of the `e` message, the server MUST verify the list of certificate hashes. If the list is empty, it MUST fail the handshake. For every certificate in the list, it checks if it possesses a certificate with the corresponding hash. If so, it continues with the handshake. However, if there is even a single certificate hash in the list that it cannot associate with a certificate, it MUST abort the handshake.
+
+For the client, the libp2p connection is fully established once it has sent the last Noise handshake message. For the server, receipt (and successful verification) of that message completes the handshake.
```

Comment:
> For the client, the libp2p connection is fully established once it has sent the last Noise handshake message.

Are the following assumptions correct?

- The Noise XX handshake has 3 messages (client -> server, server -> client, client -> server).
- The 3rd message contains the server's certificate hash send from the client to the server?
- The client depends on the server terminating the connection in case the certificate hash send by the client does not match one of its advertised certificate hashes, thus implying a MITM?

If the above is correct, would this handshake not be prone to a race condition, potentially making it insecure?

Say there is a MITM on a WebTransport connection between a client and a server.

1. The client sends the 3rd message of the Noise handshake containing the MITM certificate hash, not the correct server certificate hash.
2. The client assumes the connection to be secure, opens a second WebTransport channel and sends user-data on that channel.
3. The server notices the certificate hash mismatch and aborts the handshake.
4. The client notices that the server aborted the handshake.

Given that (2) happens before (4) the client sends user-data on a channel that is MITM.

Reply:
> The 3rd message contains the server's certificate hash send from the client to the server.

This assumption is not correct. The certificate hash is sent on the first message. This message is neither encrypted nor authenticated (at that point), but it is part of the rolling hash, which means it's authenticated retroactively: the handshake will fail if an attacker modifies it.

At least that's my understanding of the Noise handshake, unfortunately the spec is not really clear about this. Here's the Go code: https://github.com/flynn/noise/blob/d803f5c4b0f4c9bb4d8234603b9cf6c194f6b135/state.go#L443-L446